TypeScript (#1819)
enkidevs/curriculum: typescript/introduction/tsc-interfaces-classes/tsc-classes-intro.md

---
author: kapnobatai136

type: normal

category: must-know

aspects:
  - introduction

---

# Classes Intro

---
## Content

> Classes represent the building blocks of the style of programming called *object-oriented (class-based) programming*, or OOP[1]. 

Reviewer: This statement doesn't seem right. object-oriented is not exactly *class based* but rather making classes is a common way of implementing OO paradigm. JS uses functions (before ES6) to do OO

Reviewer 2: So what this is supposed to say is that this is [class-based OOP](https://en.wikipedia.org/wiki/Class-based_programming).

JS uses [prototype-based](https://en.wikipedia.org/wiki/Prototype-based_programming) OOP.

Reviewer 2: You're right that this statement isn't clear. How about just this @sourabhv?

```suggestion
Classes represent the building blocks of the style of programming called *object-oriented programming*, or OOP[1]. 
```

> The idea behind it is to organize sections of the code into definitions of entities that exist in your program (an "Animal", a "GamePlayer", a "User", etc). Then you can use these definitions, called classes, to *create (or instantiate)* those entities wherever you'd like. These instances of classes are called objects.

Reviewer 2: ```suggestion
The idea behind it is to organize related values and behaviors into definitions of entities that exist in your program (an "Animal", a "GamePlayer", a "User", etc). Then you can use these definitions, called classes, to *create (or instantiate)* those entities wherever you'd like. These instances of classes are called objects.
```

> The idea behind it is to organize related values and behaviors into definitions of entities that exist in your program (an "Animal", a "GamePlayer", a "User", etc). Then you can use these definitions, called classes, to *create (or instantiate)* those entities wherever you'd like. These instances of classes are called objects.

Reviewer: Maybe add all 4 OO concepts - Abstraction, Encapsulation, Inheritance, Polymorphism?

Reviewer 2: Hmmm maybe in the footnote (it's not beginner-friendly at this point)

> To define a class, TypeScript gives us the `class` keyword. To create an instance of that class we can use the `new` keyword. Let's look at an example:

Reviewer 3: I think this is JS syntax that's extended by typescript to allow adding types to all class variables

seems we present it like "oh this is something new in and specific to TS"

Reviewer 2: The thing is, not all people learning TypeScript already know JavaScript. Someone might encounter all of this the first time in TypeScript.

Reviewer 3: same can be said about array, floats etc. my point is we shouldn't say classes are an inherent ts feature

```ts
// define an Animal

> class Animal {
>   name: string;
>   constructor(animalName: string) {
>     this.greeting = animalName;
>   }
>   sayHi() {
>     console.log("Hi, I'm a " + this.greeting);
>   }
> }

Reviewer: ```suggestion
class Animal {
  name: string;
  constructor(animalName: string) {
    this.name = animalName;
  }
  sayHi() {
    console.log("Hi, I'm a " + this.name);
  }
}
```

// create (instantiate) an Animal
let zebra = new Animal("zebra");
zebra.sayHi();
// Hi, I'm a zebra
```

The benefit of classes is that they allow us to define data and behavior in one place, and then create many entities that use that data and follow that behavior as we see fit. In the example above, we used the `Animal` class to create a `zebra` but we could've also created a `tiger` and an `eagle` too. This allows us to have reusable code and not repeat ourselves.

> Note that using classes comes with some drawbacks as well[2]

Reviewer 3: let's add a tiny example - "...such as .."

Reviewer 2: The example is in the footnote

Reviewer 3: exactly. without the footnote (which should be optional) the statement doesn't say too much

Reviewer 2: So do you think we should move what's in the footnote here?

Or something like this:

```suggestion
Note that using classes comes with some drawbacks as well due to their strict rules[2]
```

---
## Practice

Create an instance of the `City` class:

```ts
class City {
  name: string;
  constructor(cityName: string) {
    this.name = cityName;
  }
  printCity() {
    console.log(this.name);
  }
}

let london = ??? ???('London')
```

* new
* City
* New
* city

---
## Footnotes

[1:Object-oriented programming] 

> OOP teaches us to organize data along with its associated behaviors into units called classes. Other classes can inherit this data/behaviors further and add their own.

Reviewer 2: @sourabhv, related to [your comment](https://github.com/enkidevs/curriculum/pull/1819/files/78e43d2ba6a665799a5199ef305b2f52aa85103a#r415727153), how about we do this?

```suggestion
OOP teaches us to [organize](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)) data along with its associated behaviors into [units](https://en.wikipedia.org/wiki/Abstraction_(computer_science)) called classes. Other classes can [inherit](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)) this data/behaviors further and add their own and [related classes can be converted into one another](https://en.wikipedia.org/wiki/Polymorphism_(computer_science)).
```

Reviewer: Looks good but `associated behaviors into [units] (https://en.wikipedia.org/wiki/Abstraction_(computer_science)`  but does this mean  abstraction? Also ~~this data/behaviors further and add their own~~ this data/behaviors further and add their own or change existing ones.
